src/react/utils.js

```diff
@@ -58,6 +60,92 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
   return false;
 }
 
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
+    } else if (prop instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (prop instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+function canHoistArray(realm: Realm, array: ArrayValue, hoistedScope?: Generator): boolean {
+  let lengthValue = Get(realm, array, "length");
+  invariant(lengthValue instanceof NumberValue);
+  let length = lengthValue.value;
+  for (let i = 0; i < length; i++) {
+    let element = Get(realm, array, "" + i);
+
+    if (element instanceof AbstractValue) {
+      return false;
+    } else if (element instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, element, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (element instanceof FunctionValue && element.parent !== hoistedScope) {
+      return false;
+    } else if (element instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, element, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+export function canHoistReactElement(realm: Realm, reactElement: ObjectValue, hoistedScope?: Generator): boolean {
+  if (reactElement.$CanHoist) {
+    return true;
+  }
+  // we host components <div />, <span /> etc
+  // we also allow functions where their parent scope matches the hoisted scope
```

Comment:
What does hoisting a function into its parent scope mean? Isn't a function already declared in its parent scope?

Reply:
A function might be declared in the additionalFunction, and be inaccessible in the parent scope.